# Проект спринта 7 - Развитие игровой индустрии

- Автор: Алёна Опольская
- Дата: 7 января 2026

### Цели и задачи проекта

<font color='#777778'> В этой тетрадке используем датасет new_games.csv, который содержит информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр. Задача проекта - помочь коллегам из команды игры «Секреты Темнолесья», а именно познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных, что будет использовано коллегами для дальнейшей подготовки статьи-исследования о развитии индустрии игр в начале XXI века. </font>

### Описание данных

<font color='#777778'>
    В проекте будут использованы данные датасета new_games.csv (коллеги собрали данные из открытых источников) с таким описанием:
- `Name` — название игры;
- `Platform` название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.
</font>

### Содержимое проекта
<font color='#777778'>
1. Загрузка данных и знакомство с ними <br>
2. Проверка ошибок в данных и их предобработка <br>
    2.1 Названия столбцов датафрейма <br>
    2.2 Типы данных <br>
    2.3 Наличие пропусков в данных <br>
    2.4 Явные и неявные дубликаты в данных <br>
3. Фильтрация данных <br>
4. Категоризация данных <br>
5. Итоговый вывод </font>

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные из датасета new_games.csv в датафрейм games
games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
# Выводим первые 5 строк датасета
games.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [4]:
# Выводим информацию о данных
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Данные датасета содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр. Видя первые 5 строк, можно сказать о том, что данные соответствуют описанию, полученному в задании. Всего нам представлены данные объемом 11 столбцов и 16956 строк.

Стоит отметить, что в данных есть пропуски, а именно в следующих столбцах: Name, Year of Release,  Genre, Critic Score, User Score, Rating. В столбцах Name и Genre всего по 2 пропуска, а вот остальные столбцы требуют большего внимания и обработки перед анализом: в Year of Release 275 пропусков, в Critic Score - 8714, в User Score - 6804, а в Rating - 6871. 

Стоит обратить внимание на тип данных, ведь в некоторых столбцах данные представлены не в самом удобном виде:
    год выпуска игры должен быть целым, а сейчас представлен в виде дробного числа;
    продажи в Европе и продажи в Японии должны быть выражены числами (в миллионах проданных копий), а сейчас имеют тип object;
    то же самое со столбцом User Score - согласно описанию данных, это оценка от 0 до 10, т.е. число, а сейчас данные типа object.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [5]:
# Выводим названия всех столбцов датафрейма
games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
# Приводим названия столбцов к нужному стилю написания
columns_snake_case = (games.columns.str.lower() .str.replace(' ','_'))

In [7]:
# Выводим названия столбцов
games.columns = columns_snake_case
games.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

In [8]:
# Комментарий ревьюера
# создаем копию датасета до преобразования для возможности проверить сделанные изменения после предобработки
tmp = games.copy() 
len(tmp)

16956

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [9]:
# Проверим уникальные значения столбца eu_sales
games['eu_sales'].unique()

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

In [10]:
# Проверим уникальные значения столбца jp_sales
games['jp_sales'].unique()

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

В столбцах eu_sales и jp_sales числа представлены строковым типом (о чем нам говорят кавычки ''), а еще есть значение 'unknown'. Поэтому данные имеют тип object вместо числового типа.

In [11]:
# Проверим уникальные значения столбца year_of_release
games['year_of_release'].unique()

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

В столбце year_of_release есть пропуски (nan), поэтому сейчас он имеет не целочисленный тип. 

In [12]:
# Проверим уникальные значения столбца user_score
games['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

В столбце user_score есть пропуски, данные 'tbd' и числа представлены типом str (снова обращаем внимание на кавычки), поэтому сейчас он имеет тип object вместо числового.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [13]:
# Приведем к числовому типу данные столбцов eu_sales, jp_sales и user_score (строковые значения unknown или другие заменим на пропуски с помощью аргумента 'coerce' )
games['eu_sales'] = pd.to_numeric(games['eu_sales'], errors='coerce')
games['jp_sales'] = pd.to_numeric(games['jp_sales'], errors='coerce')
games['user_score'] = pd.to_numeric(games['user_score'], errors='coerce')

In [14]:
# Приведем данные года выпуска игры к типу Int64 (не удаляем строки и сохраняем пропуски)
games['year_of_release'] = games['year_of_release'].astype('Int64')

In [15]:
# Выводим информацию о данных для проверки типов данных в столбцах
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  Int64  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


In [16]:
# Выводим первые 5 строк датасета для проверки отображения данных
games.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E
1,Super Mario Bros.,NES,1985,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E
4,Pokemon Red/Pokemon Blue,GB,1996,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [17]:
# Считаем количество пропусков в каждом столбце
games.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [18]:
# Считаем долю пропусков в столбцах датафрейма
games.isna().mean()

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

In [20]:
# Считаем долю пропусков в столбцах датафрейма в %
games.isna().sum() / len(games) * 100 

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


Стоит обратить внимание на то, что количество пропусков изменилось в большую сторону (по сравнению с пераичной проверкой методом info). 

Это связано с тем, что в ходе изменения типа данных мы заменяли строковые значения (unknown, tbd и др) на пропуски. Больше всего пропусков в стобцах с оценками критиков и пользователей, а также в рейтинге организации ESRB. Это может быть связано с:
    нехватка данных (пользователи, которые покупали игру, просто не оставляли отзывов в виде оценки);
    можно предположить, что вместо tbd позже должны были добавить оценку, но этого еще не было сделано;
    для некоторых старых игр не собирался рейтинг (тот же рейтинг организации ESRB был разаротан в 1994 году) или данные могли быть утеряны/не предоставлены.

Пропусков в остальных столбцах гораздо меньше: год выпуска игры (275 пропусков) может быть не известен; название игры и жанр (по 2 пропуска) - скорее всего просто технические ошибки при сборе данных; а пропуске в столбцах "продажи в Японии" и "продажи в Европе" обусловлены изменением типа данных (мы заменяли unknown на пропуски).

В задании сказано, что коллег интересует развитие игровой индустрии с 2000 по 2013 год. Проверим какого года выпуска игры, название и жанр которых отсутствуют.

In [21]:
# Выводим строки с пропусками name
print(f'Вот строки с пропусками в столбце name:')
print(games[games['name'].isna()].to_string(index=False))
# Выводим строки с пропусками genre
print(f'Вот строки с пропусками в столбце genre:')
print(games[games['genre'].isna()].to_string(index=False))

Вот строки с пропусками в столбце name:
name platform  year_of_release genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score rating
 NaN      GEN             1993   NaN      1.78      0.53      0.00         0.08           NaN         NaN    NaN
 NaN      GEN             1993   NaN      0.00      0.00      0.03         0.00           NaN         NaN    NaN
Вот строки с пропусками в столбце genre:
name platform  year_of_release genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score rating
 NaN      GEN             1993   NaN      1.78      0.53      0.00         0.08           NaN         NaN    NaN
 NaN      GEN             1993   NaN      0.00      0.00      0.03         0.00           NaN         NaN    NaN


Мы видим, что год выпуска 1993, а значит эти 2 строки никак не будут влиять на наш анализ, т к в дальнешей выборке мы будем использовать игры 2000-2013 года, поэтому эти строки можно или не трогать, или удалить (вторая строка вообще странная - отсутствуб какие либо данные кроме jp_sales, ее точно можно удалить без потерь). 

Строки с пропусками в столбце "год выпуска игры" не требуют удаления, т к они автоматически не войдут в анализ, когда мы дойдем до выборки данных по интересующим нас годам.

Пропуски в данных "продажи в Японии" и "продажи в Европе" можно заменить на средние значения, т к их не много и они не окажут значительного влияния на результаты анализа (не исказят результаты).

Пропуски данных в "рейтинг" не трогаем, т к в дальнейшем задании по категоризации эти данные не будут учитываться. А вот с пропусками в оценках пользователей и критиков сложнее - именно по ним будет проводиться категоризация. Для принятия решения о том, что делать с пропусками, рассмотрим как пропуски распределяются по годам.

In [22]:
# Группируем пропуски critic_score по годам и считаем количество 
no_critic_grouped_year = (
    games[games['critic_score'].isna()]
    .groupby('year_of_release')['name']
    .count()
)
print(f'Количество пропусков в столбце critic_score:')
print(no_critic_grouped_year)
# Группируем пропуски user_score по годам и считаем количество 
no_user_grouped_year = (
    games[games['user_score'].isna()]
    .groupby('year_of_release')['name']
    .count()
)
print(f'Количество пропусков в столбце user_score:')
print(no_user_grouped_year)

Количество пропусков в столбце critic_score:
year_of_release
1980      9
1981     46
1982     37
1983     18
1984     14
1985     13
1986     22
1987     17
1988     14
1989     17
1990     16
1991     42
1992     42
1993     60
1994    120
1995    220
1996    259
1997    276
1998    356
1999    301
2000    213
2001    158
2002    205
2003    194
2004    204
2005    288
2006    394
2007    511
2008    720
2009    791
2010    770
2011    646
2012    345
2013    274
2014    322
2015    385
2016    275
Name: name, dtype: int64
Количество пропусков в столбце user_score:
year_of_release
1980      9
1981     46
1982     37
1983     18
1984     14
1985     13
1986     22
1987     17
1988     14
1989     17
1990     16
1991     42
1992     42
1993     60
1994    120
1995    220
1996    259
1997    275
1998    353
1999    309
2000    250
2001    228
2002    370
2003    264
2004    278
2005    360
2006    464
2007    578
2008    775
2009    825
2010    802
2011    645
2012    326
2013    243
201

В резуьтате мы видим, что на интересующий нас период (2000-2013 гг) приходится большое количество пропусков, поэтому заменять пропуски средним значением не корректно в виду необходимости дальнейшей категоризации игр по оценкам. Это сильно повлияет на резуьтаты категоризации и анализа, в связи с чем решаем оставить данные как есть и в дальнейшем при категоризации выделять отдельную группу, где отсутствуют нужные данные.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

Две строки, в которых отсутствуют название и жанр, можно удалить - по факту с данными там вообще одна строка, и то, в ней присутствуют только год выпуска и количество продаж, а название игры, жанр, рейтинг и оценки отсутствуют - это не информативно для анализа и удаление не повлияет на результаты, т к это всего одна строка.

In [23]:
# Удаляем строки с пропусками в name и genre и снова проверяем количество пропусков
games = games.dropna(subset=['name', 'genre'])
print(games[['name', 'genre']].isna().sum())

name     0
genre    0
dtype: int64


Строки с пропусками в столбце "год выпуска игры" не трогаем, т к они автоматически не войдут в анализ, когда мы дойдем до выборки данных по интересующим нас годам. Также не трогаем пропуски данных в "рейтинг".

Пропуски в данных "продажи в Японии" и "продажи в Европе" можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [24]:
# Заменяем пропуски в продажах в Европе и в Японии на средние значения в зависимоит от платформы и года выходы игры
games['eu_sales'] = games['eu_sales'].fillna(
    games.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean')
)

games['jp_sales'] = games['jp_sales'].fillna(
    games.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean')
)
# Проверяем количество пропусков
print(games[['eu_sales', 'jp_sales']].isna().sum())

eu_sales    0
jp_sales    0
dtype: int64


Пропуски в user_score и critic_score мы не удаляем и не заменяем (причины описаны выше), далее при категоризации мы будем выделять их в отдельную группу, чтобы не искажать результаты анализа.

In [25]:
# Комментарий ревьюера
# Посмотрим какие пропуски остались
show_missing_stats(games)

,Кол-во пропусков,Доля пропусков
year_of_release,275,0.0162
critic_score,8712,0.5139
user_score,9266,0.5465
rating,6869,0.4052


In [26]:
# Удаляем строки с пропусками в "год выпуска игры" и снова проверяем количество пропусков
games = games.dropna(subset=['year_of_release'])
print(games[['year_of_release']].isna().sum())

year_of_release    0
dtype: int64


In [27]:
# Посмотрим какие пропуски остались
show_missing_stats(games)

,Кол-во пропусков,Доля пропусков
critic_score,8594,0.5153
user_score,9121,0.5469
rating,6778,0.4064


In [28]:
# Заполним пропуски в рейтинге ESRB "no_esrb"
games['rating'] = games['rating'].fillna('no_esrb')
# Посмотрим какие пропуски остались
show_missing_stats(games)

,Кол-во пропусков,Доля пропусков
critic_score,8594,0.5153
user_score,9121,0.5469


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [29]:
# Посчитаем количество вхождений каждого уникального значения в столбцах
print(games['genre'].value_counts())
print(games['platform'].value_counts())
print(games['rating'].value_counts())
print(games['year_of_release'].value_counts())

Action          3341
Sports          2325
Misc            1740
Role-Playing    1493
Shooter         1314
Adventure       1309
Racing          1244
Platform         891
Simulation       866
Fighting         844
Strategy         679
Puzzle           577
ACTION            12
SPORTS             7
FIGHTING           6
RACING             6
ROLE-PLAYING       6
SHOOTER            5
ADVENTURE          4
PLATFORM           3
MISC               3
PUZZLE             2
SIMULATION         1
STRATEGY           1
Name: genre, dtype: int64
PS2     2154
DS      2147
PS3     1330
Wii     1305
X360    1249
PSP     1212
PS      1208
PC       972
GBA      826
XB       818
GC       549
3DS      522
PSV      434
PS4      395
N64      320
XOne     251
SNES     241
SAT      174
WiiU     147
2600     118
NES      100
GB        97
DC        52
GEN       27
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: platform, dtype: int64
no_esrb    6778
E          3968
T     

In [30]:
# Исправим дубликаты в жанрах (написаны разным регистром, приведем все к нижнему и проверим)
games['genre'] = games['genre'].str.lower()
print(games['genre'].value_counts())

action          3353
sports          2332
misc            1743
role-playing    1499
shooter         1319
adventure       1313
racing          1250
platform         894
simulation       867
fighting         850
strategy         680
puzzle           579
Name: genre, dtype: int64


Больше ничего не исправляем.

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [31]:
# Посчитаем количество явных дубликатов, если они есть
games.duplicated().sum()

235

In [32]:
# Определяем дубликаты, раз они есть и выведем первые 6, чтобы проверить, что это полные дубликаты
dupli = games[games.duplicated(keep=False)]
dupli.head(6)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
267,Batman: Arkham Asylum,PS3,2009,action,2.24,1.31,0.07,0.61,91.0,8.9,T
268,Batman: Arkham Asylum,PS3,2009,action,2.24,1.31,0.07,0.61,91.0,8.9,T
367,James Bond 007: Agent Under Fire,PS2,2001,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T
368,James Bond 007: Agent Under Fire,PS2,2001,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T
716,God of War: Ascension,PS3,2013,action,1.23,0.63,0.04,0.35,80.0,7.5,M
717,God of War: Ascension,PS3,2013,action,1.23,0.63,0.04,0.35,80.0,7.5,M


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

Как видно выше, в датасете найден 235 явный дубликат. Вывели первые 6 строк (с keep=False, чтобы увидеть пары дубликатов) и убедились, что дубликаты полные - совпадения по всем столбцам. Для дальнейшего корректного анализа такие дубликаты нужно удалить, они не несут в себе смысла.

In [33]:
# Удалим явные дубликаты и проверим новое количество дубликатов
games = games.drop_duplicates()
games.duplicated().sum()

0

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

Посчитаем удаленные строки и проверим новое количество строк. 

In [34]:
# Считаем количество удаленных строк (в абсолюте и в процентах)
delete_rows = len(tmp) - len(games)
percent_delete_rows = delete_rows/len(tmp)*100
print(f'Количество удаленных строк: {delete_rows}')
print(f'Количество удаленных строк в %: {round(percent_delete_rows,2)}')
print(f'Новое количество строк: {len(games)}')

Количество удаленных строк: 512
Количество удаленных строк в %: 3.02
Новое количество строк: 16444


- После проведения предобработки данных напишите общий промежуточный вывод.

Мы познакомились с данными о играх разных жанров и платформ. Провели предобработку данных, в ходе которой:
- удалили строки с отсутствующими данными name и genre (2 строки);
- <font color='Blue'><b>удалили строки с отсутствующим годом выхода игры;
- заменили пропуски в рейтинге на no_esrb;</b></font>
- заполнили пропуски в столбцах eu_sales и jp_sales путем расчета средних значений с учетом года выпуска игры и платформы;
- приняли решение не заполнять средним отсутствущие значения в оценках, чтобы избежать искажения результатов анализа при дальнейшей категоризации;
- выявили неявные дубликаты и исправили написание жанров (привели к нижнему регистру);
- выявили явные дубликаты и удалили их (241 строка).

Теперь датасет готов для для дальнейшего анализа.

In [35]:
# Выводим информацию о данных после предобработки
games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16444 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  Int64  
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     7983 non-null   float64
 9   user_score       7463 non-null   float64
 10  rating           16444 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.5+ MB


---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [36]:
# Отберем данные периода с 200 по 2013 год включительно
df_actual = games[
    (games['year_of_release'] >= 2000) &
    (games['year_of_release'] <= 2013)
]

In [37]:
# Посмотрим инфо нового датасета с учетом фильтрации по периоду, а также проверим минимальное и максимальное значения годов
min_year = df_actual['year_of_release'].min()
max_year = df_actual['year_of_release'].max()
df_actual.info()
print(f'Минимальный год: {min_year}')
print(f'Максимальный год: {max_year}')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12781 non-null  object 
 1   platform         12781 non-null  object 
 2   year_of_release  12781 non-null  Int64  
 3   genre            12781 non-null  object 
 4   na_sales         12781 non-null  float64
 5   eu_sales         12781 non-null  float64
 6   jp_sales         12781 non-null  float64
 7   other_sales      12781 non-null  float64
 8   critic_score     7169 non-null   float64
 9   user_score       6483 non-null   float64
 10  rating           12781 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.2+ MB
Минимальный год: 2000
Максимальный год: 2013


<font color='DarkBlue'><b>Комментарий ревьюера</b></font><br>
<font color='DarkGreen'>👌 Правильно отфильтровали данные по годам релиза. А проверку результата фильтрации можно провести еще вот так:</font>

In [38]:
# Комментарий ревьюера
df_actual.year_of_release.sort_values().unique()

<IntegerArray>
[2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012,
 2013]
Length: 14, dtype: Int64

После фильтрации по периоду в датасете осталось так же 11 столбцов, но 12781 строка. По мин и макс годам видим, что период выбран правильно. Пропуски присутствуют только в оценках и рейтинге.

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [39]:
# Проводим категоризацию по оценкам пользователей
df_actual.loc[:, 'user_score_cat'] = pd.cut(
    df_actual['user_score'], 
    bins=[0, 3, 8, 10], 
    labels=['низкая оценка', 'средняя оценка', 'высокая оценка'], 
    right=False
)

df_actual.loc[:, 'user_score_cat'] = (
    df_actual['user_score_cat']
    .cat.add_categories('без оценки')
    .fillna('без оценки')
)

/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1597: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = value
/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [40]:
# Проводим категоризацию по оценкам критиков
df_actual.loc[:, 'critic_score_cat'] = pd.cut(
    df_actual['critic_score'], 
    bins=[0, 30, 80, 100], 
    labels=['низкая оценка', 'средняя оценка', 'высокая оценка'], 
    right=False
)

df_actual.loc[:, 'critic_score_cat'] = (
    df_actual['critic_score_cat']
    .cat.add_categories('без оценки')
    .fillna('без оценки')
)

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [41]:
# Посмотрим количество игр в каждой категории по user_score, учитывая отсутствующие данные
user_score_cat_counts = df_actual['user_score_cat'].value_counts(dropna=False)
user_score_cat_share = user_score_cat_counts / len(df_actual) * 100
print('Количество игр в категориях:')
print(user_score_cat_counts)
print('Количество игр в категориях в %:')
print(round(user_score_cat_share,2))   
print(f'Общее количество игр: {len(df_actual)}')

Количество игр в категориях:
без оценки        6298
средняя оценка    4081
высокая оценка    2286
низкая оценка      116
Name: user_score_cat, dtype: int64
Количество игр в категориях в %:
без оценки        49.28
средняя оценка    31.93
высокая оценка    17.89
низкая оценка      0.91
Name: user_score_cat, dtype: float64
Общее количество игр: 12781


In [42]:
# Посмотрим количество игр в каждой категории по critic_score, учитывая отсутствующие данные
critic_score_cat_counts = df_actual['critic_score_cat'].value_counts(dropna=False)
critic_score_cat_share = critic_score_cat_counts / len(df_actual) * 100
print('Количество игр в категориях:')
print(critic_score_cat_counts)
print('Количество игр в категориях в %:')
print(round(critic_score_cat_share,2))   
print(f'Общее количество игр: {len(df_actual)}')

Количество игр в категориях:
без оценки        5612
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: critic_score_cat, dtype: int64
Количество игр в категориях в %:
без оценки        43.91
средняя оценка    42.42
высокая оценка    13.24
низкая оценка      0.43
Name: critic_score_cat, dtype: float64
Общее количество игр: 12781


- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [43]:
# Считаем количество игр на каждой платформе и выделяем топ-7
games_on_platform = df_actual['platform'].value_counts()
top_platforms = games_on_platform.head(7)
print('Топ-7 платформ по количеству игр (2000-2013 гг):')
print(top_platforms)

Топ-7 платформ по количеству игр (2000-2013 гг):
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: platform, dtype: int64


---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Мы провели анализ данных о видеоиграх в несколько этапов: познакомились с данными, проверили их корректность, провели предобработку и получили необходимый срез данных.

На первом этапе мы загрузили данные и ознакомились с ними. Нам были предоставлены данные объемом 11 столбцов и 16956 строк. Было замечено, что в данных есть пропуски, а также накоторые данные представлены некорретным типом. 

На следующем этапе мы углубились в изучение ошибок в данных и предобработку: привели названия всех столбцов к стилю snake case, привели все данные к корректным типам, заменили не понятные строковые значения в числовых данных на пропуски. 

Далее работали с пропусками в данных: 
- было решено удалить строки с пропусками в столбцах name и genre; 
- заполнить средними значениями, рассчитанными по платформе и году выпуска игры, пропуски в количестве продаж (eu_sales и jp_sales); 
- пропуски в столбце года выпуска игры удалили; 
- пропуски в рейтинге заменили на 'no_ersb';
- пропуски в оценках пользователей и критиков были оставлены без заполнения средним значением, так как их доля высока и заполнение могло бы исказить результаты анализа.

Затем мы проверили данные на наличие неявных и явных дубликатов:
- названия жанров игр были написаны разными регистрами, мы привели их к нижнему регистру;
- выявили и удалили строки с полным совпадением данных.

В итоге после предобработки в датасете осталось 16444 строк.

Далее мы отобрали данные с 2000 по 2013 год включительно и сохранили этот срез в датафрейме df_actual. Таких строк оказалось 12781. После этого мы разделили игры на нужные категории:

1) категоризация по оценкам пользователей: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала). Среди игр, по которым есть данные user_score, наибольшая категория - игры со средней оценкой (примерно 32%). Далее распределение такое: высокая оценка 17.89%, низкая оценка 0.91%;

2) категоризация по оценкам критиков: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала). Среди игр, по которым есть данные critic_score, наибольшая категория - игры со средней оценкой (около 42%). Высокая оценка   - 13.24%, низкая оценка - 0.43%.

Важно отметить, что в обоих категоризациях высока доля отсутствующих данных: 49% и 43% соответственно. Мы вынесли их в отдельную категорию и не заменяли средним значением, с целью избежания искажения результатов.

Далее были выделены топ-7 платформ по количеству игр, выпущенных за весь актуальный период: PS2, DS, Wii, PSP, X360, PS3, GBA.